EventListener/ResponseStatusCodeListener.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+/*
+ * This file is part of the FOSRestBundle package.
+ *
+ * (c) FriendsOfSymfony <http://friendsofsymfony.github.com/>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace FOS\RestBundle\EventListener;
+
+use FOS\RestBundle\FOSRestBundle;
+use FOS\RestBundle\Util\ExceptionValueMap;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpKernel\Event\ExceptionEvent;
+use Symfony\Component\HttpKernel\Event\ResponseEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+
+/**
+ * @author Christian Flothmann <christian.flothmann@sensiolabs.de>
+ */
+class ResponseStatusCodeListener implements EventSubscriberInterface
+{
+    private $exceptionValueMap;
+
+    public function __construct(ExceptionValueMap $exceptionValueMap)
+    {
+        $this->exceptionValueMap = $exceptionValueMap;
+    }
+
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            KernelEvents::EXCEPTION => 'getResponseStatusCodeFromThrowable',
+            KernelEvents::RESPONSE => 'setResponseStatusCode',
+        ];
+    }
+
+    /**
+     * @param ExceptionEvent $event
+     */
+    public function getResponseStatusCodeFromThrowable($event): void
+    {
+        $request = $event->getRequest();
+
+        if (!$request->attributes->get(FOSRestBundle::ZONE_ATTRIBUTE, true)) {
+            return;
+        }
+
+        if (method_exists($event, 'getThrowable')) {
+            $throwable = $event->getThrowable();
+        } else {
+            $throwable = $event->getException();
+        }
+
+        $statusCode = $this->exceptionValueMap->resolveThrowable($throwable);
+
+        if (is_int($statusCode)) {
+            $request->attributes->set(FOSRestBundle::STATUS_CODE_ATTRIBUTE, $statusCode);
```

Comment:
I think we could just set a private property on the ResponseStatusCodeListener instead of polluting the request class.

Reply:
agree with Tabion

Reply:
Not sure that's a good idea as it could easily break with sub requests.

Reply:
I think we should skip sub-request here. I don't see the value for subrequest. Also the  old logic with the ExceptionController was also only called for the master request.

Reply:
Fair enough, I am still not a big fan of making the service stateful though, but if turns out not always work as desired we can switch to the initial implementation without any issues. I have also added a test to ensure that the listener only acts on master requests.